### ISSUE TITLE
unused_unit causes syntax error in weird loop code

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9946]
### Summary

.

### Lint Name

unused_unit

### Reproducer

I tried this code:

```rust
// run-pass
#![allow(unreachable_code)]
#![allow(for_loops_over_fallibles)]
#![deny(unused_variables)]

fn main() {
    for _ in match return () {
        () => Some(0),
    } {}
}
```

I saw this happen:
`cargo clippy --fix  --  -Wclippy::pedantic  -Aclippy::match_single_binding`
```
after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error: expected one of `.`, `;`, `?`, `}`, or an operator, found `=>`
 --> src/main.rs:8:12
  |
8 |         () => Some(0),
  |            ^^ expected one of `.`, `;`, `?`, `}`, or an operator

error: expected `{`, found `}`
  --> src/main.rs:10:1
   |
10 | }
   | ^ expected `{`

error: aborting due to 2 previous errors

Original diagnostics will follow.

warning: unneeded `()`
 --> src/main.rs:7:27
  |
7 |     for _ in match return () {
  |                           ^^ help: remove the `()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unused_unit
  = note: `#[warn(clippy::unused_unit)]` on by default
```

I expected to see this happen:

### Version

```text
rustc 1.67.0-nightly (b3bc6bf31 2022-11-24)
binary: rustc
commit-hash: b3bc6bf31265ac10946a0832092dbcedf9b26805
commit-date: 2022-11-24
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
